jmh-benchmarks/README.md

```diff
@@ -0,0 +1,196 @@
+###JMH-Benchmark module
+
+This module is used to contain benchmarks written using [JMH](http://openjdk.java.net/projects/code-tools/jmh/)
+from openjdk. Writing correct micro-benchmarks is Java (or another JVM language) is difficult
+and there are many non-obvious pitfalls (many due to compiler optimizations).  JMH is a framework for running
+and analyzing benchmarks (micro or macro) written in Java (or another JVM language). 
+
+For help in writing correct JMH tests, the best place to start is the [sample code](http://hg.openjdk.java.net/code-tools/jmh/file/tip/jmh-samples/src/main/java/org/openjdk/jmh/samples/) provided
+by the JMH project.
+
+Typically, JMH is expected to run as a separate project in Maven.  The jmh-benchmarks module uses
+the [gradle shadow jar](https://github.com/johnrengelman/shadow) plugin to emulate this behavior, by creating the required
+uber-jar file containing the benchmarking code and required JMH classes.  
+
+JMH is highly configurable and users are encouraged to look through the samples for suggestions
+on what options are available. A good tutorial for using JMH can be found [here](http://tutorials.jenkov.com/java-performance/jmh.html#return-value-from-benchmark-method)
+
+###Gradle Tasks / Running benchmarks in gradle
+
+If no benchmark mode is specified, the default is used which is throughput.
+
+*  jmh-benchmarks:shadowJar - creates the uber jar required to run the benchmarks.
+
+*  jmh-benchmarks:jmh - runs the clean, shadowJar tasks then runs all the benchmarks. 
+ 
+*  To run a specific benchmark (MemoryBytesCacheBenchmark.java for example): 
+        `./gradlew jmh-benchmarks:jmh -DjmhArgs=MemoryBytesCacheBenchmark` 
+
+*  To specify jmh specific flags, just add the desired flag and value (again no classname being specified runs all benchmarksHere).  Here
+     is an example for running the benchmark class (MemoryBytesCacheBenchmark.java) setting the "forks" property:
+          `./gradlew jmh-benchmarks:jmh -DjmhArgs=-f2,MemoryBytesCacheBenchmark`.  
+     The -f flag in JMH is the number of "forks" or test runs to do, with the default setting being 10.  Each fork or test run does 20 warmup iterations (default) and 20 benchmarking
+     iterations (again a default value).
+
+* By default all output goes to stdout.  If it is desired to capture the output in a file add -o&lt;filename&gt; to the jmhArgs property:
+   `./gradlew jmh-benchmarks:jmh -DjmhArgs=-f2,-oMemBytesBenchmark.txt,MemoryBytesCacheBenchmark`
```

Comment:
@ijuma  thanks for the comments.

 I agree using -DjmhArgs is clunky.  But in the gradle-jmh-plugin all JMH parameters are configured in a static block in the build.gradle file.  If you want/need different configs for individual benchmarks then variables in that block need to be changed.  (It seems that command line options in the jmh-gradle-plugin are an unresolved issue https://github.com/melix/jmh-gradle-plugin/issues/46, https://github.com/melix/jmh-gradle-plugin/issues/54)

I have an alternante proposal.  What do you think about a shell script that:

1. Accepts plain JMH arguments;  "./jmh -o outPutfile.txt -f 10 MyBenchmarks"
2. Delegates to gradle to build the shadow jar file.
3. Runs JMH from command line with args passed in step 1.